zilliqa/src/bin/validator_oracle.rs

```diff
@@ -0,0 +1,238 @@
+use std::{path::PathBuf, str::FromStr};
+
+use alloy_provider::{Provider, ProviderBuilder, WsConnect};
+use alloy_rpc_types::{BlockNumberOrTag, Filter};
+use anyhow::Result;
+use clap::Parser;
+use ethers::{
+    contract::abigen,
+    providers::Middleware,
+    signers::{LocalWallet, Signer},
+    types::{TransactionRequest, H160},
+};
+use futures_util::stream::StreamExt;
+use tokio::sync::watch;
+use tracing::{debug, error, info};
+use tracing_subscriber::EnvFilter;
+use zilliqa::{
+    contracts,
+    crypto::{NodePublicKey, SecretKey},
+    state::contract_addr,
+    uccb::{
+        cfg::{ChainConfig, Config, ZQ2Config},
+        client::ChainClient,
+    },
+};
+
+// The json files are generated by invoking 'forge build' in zilliqa/src/contracts/uccb
+abigen!(
+    ValidatorManager,
+    "zilliqa/src/contracts/uccb/out/ValidatorManager.sol/ValidatorManager.json"
+);
+
+#[derive(Parser, Debug)]
+struct Args {
+    #[arg(value_parser = SecretKey::from_hex)]
+    secret_key: SecretKey,
+    #[clap(long, short, default_value = "config.toml")]
+    config_file: PathBuf,
+}
+
+impl zilliqa::uccb::Args for Args {
+    fn secret_key(&self) -> &SecretKey {
+        &self.secret_key
+    }
+
+    fn config_file(&self) -> &PathBuf {
+        &self.config_file
+    }
+}
+
+struct ValidatorOracle {
+    wallet: LocalWallet,
+    zq2_config: ZQ2Config,
+    validators: Vec<NodePublicKey>,
```

Comment:
I'm confused how the synchronisation of the validator set is implemented here. We seem to have both a mutable field which gets updated, plus a `tokio::sync::watch` down which the updated set is sent too. Why do we need both?

It seems sensible to me to just have the channel approach, since we only need to send a transaction when the set is updated?

Reply:
We probably don't need both - your'e right. Initially thought this might be needed to figure out which validators have been added/removed or use internally for something else. I'll remove the internal field.